physics/continuous_trajectory_body.hpp

```diff
@@ -442,32 +446,66 @@ ContinuousTrajectory<Frame>::ReadFromMessage(
             Polynomial<Position<Frame>, Instant>::template ReadFromMessage<
                 EstrinWithoutFMA>(polynomial));
       } else {
-        serialization::Polynomial const& polynomial = pair.polynomial();
-        if (polynomial.HasExtension(
-                serialization::PolynomialInMonomialBasis::extension) &&
+        not_null<serialization::Polynomial const*> polynomial =
+            &pair.polynomial();
+        bool const is_pre_καραθεοδωρή =
+            !polynomial->HasExtension(
+                serialization::PolynomialInMonomialBasis::extension) ||
             polynomial
-                .GetExtension(
+                ->GetExtension(
                     serialization::PolynomialInMonomialBasis::extension)
-                .has_evaluator()) {
-          // The post-Καραθεοδωρή path, do not specify an evaluator when calling
-          // `ReadFromMessage`.
+                .has_evaluator();
+        if (is_pre_καραθεοδωρή) {
           continuous_trajectory->polynomials_.emplace_back(
               Instant::ReadFromMessage(pair.t_max()),
-              Polynomial<Position<Frame>, Instant>::ReadFromMessage(
-                  polynomial));
+              UseHardwareFMA ? Polynomial<Position<Frame>, Instant>::
+                                   template ReadFromMessage<Estrin>(*polynomial)
+                             : Polynomial<Position<Frame>, Instant>::
+                                   template ReadFromMessage<EstrinWithoutFMA>(
+                                       *polynomial));
         } else {
-          // The pre-Καραθεοδωρή path.
+          std::optional<serialization::Polynomial> rewritten_polynomial;
+          if (!UseHardwareFMA) {
+            // If we are on a machine without FMA, turn Estrin into
+            // EstrinWithoutFMA so that plans made on this machine can be read
+            // on a more modern machine.
+            // Note that until Lánczos, we created and serialized polynomials
+            // with Estrin even when FMA was unavailable, so we cannot trust the
+            // evaluator kind here.  Downgrading the evaluators also seems like
+            // a better way to handle the case where a save gets moved from a
+            // modern machine to an ancient one: in that direction, we cannot
+            // preserve the behaviour, but by serializing EstrinWithoutFMA at
+            // that transition, whatever work is done on the old machine will be
+            // faithfully reproduced when it is read again on a new machine.
+            auto const evaluator_kind =
+                polynomial
+                    ->GetExtension(
+                        serialization::PolynomialInMonomialBasis::extension)
+                    .evaluator()
+                    .kind();
+            if (evaluator_kind ==
+                serialization::PolynomialInMonomialBasis::Evaluator::ESTRIN) {
```

Comment:
What about `HORNER`?  We might not use it at the moment, but deserialization must be correct.